broadinstitute/single_cell_portal_core
# Easier DE upload: flexible headers, better in-app docs (SCP-5243)

This lets authors upload differential expression data much more easily, and with metrics more tailored to their research.

Here's the improved author DE upload UX:

https://github.com/broadinstitute/single_cell_portal_core/assets/1334561/2fd246ed-f4c5-4a8c-ab94-f13f2e3ce049

For comparison, here's the previous UX:

https://github.com/broadinstitute/single_cell_portal_core/assets/1334561/4badaaa6-437d-4faf-aa5b-e8bc30828c0b

### Overview
Previously (#1848), going from Scanpy or Seurat DE outputs to uploaded SCP DE results was tedious.  Users could only upload DE data that had "logfoldchanges" as the size metric and "qval" as the significance metric.  These didn't map cleanly to DE output from popular single cell analysis tools.  Column headers in the file also had to be in a strict order, and include headers for exactly "gene", "group" and "comparison_group".  No single-cell analysis tool outputs quite those headers.  The required "comparison_group" column was also often extraneous; common DE results that only have one-vs-rest comparisons don't need it.  In sum, this bespoke file format was a barrier for upload.

Now, uploading DE is much easier.  Once the user selects a file, the upload UI infers column headers from the file's first line.  The author can use the auto-populated suggestions or -- if no header is inferred or the author simply prefers a different one -- manually specify which header corresponds to which required size and significance metrics.  Headers for "gene", "group" and "comparison_group" also leverage this new convenience.  Headers can be in any order, because [Ingest Pipeline handles](https://github.com/broadinstitute/scp-ingest-pipeline/pull/322) sorting and name-mapping to output DE results files as expected by the DE explore UI.

In-app docs are also enriched and more inviting.  Users can explore a small widget to see how they can map their output from Scanpy, Seurat, or custom DE to SCP's author DE format.  For simplicity, support for wide format has been hidden; we can re-highlight it if there's demand.  Like in-app docs for other file types that need prior file uploads (e.g. processed matrices needing raw, spatial clusters needing DR clusters) the DE upload now makes that requirement implicit.  This make DE upload more welcoming to try, and easier to learn.

### Test

Existing tests have been updated to account for changes here.  To manually test new functionality:

1.  Go to upload UI for an existing study
2. Select "seurat_findallmarkers_one-vs-rest.csv" from [scp-ingest-pipeline/tests/data/author_de/](https://github.com/broadinstitute/scp-ingest-pipeline/tree/29ec077c12a62f44158d2b5a9874bc1997362a27/tests/data/author_de)
3. Confirm 5 new menus in the file panel, each with a default value shown
4. Confirm "Gene header" defaults to "gene", "Group header" to "cluster", "Comparison group header" to "N/A", "Size metric" to "avg_log2FC", and "Significance metric" to "p_val_adj".
5. Fill in values for "Associated clustering", "Associated annotation", and "Computational method"
6. Ensure `rails jobs:work` (delayed job process) is running
7. Click "Save & Upload"
8. In another terminal, run `tail -n 200 -f log/development.log`
9. Ensure job launches successfully per development.log
10. Wait 3-4 minutes
11. Ensure job completes successfully per development.log

If we get [signal that authors click on "Choose file..."](https://mixpanel.com/project/2120588/view/19059/app/boards#id=5438195&editor-card-id=%22report-44624635%22) in the "Differential expression" tab, then we should consider refinements like [clearing warnings upon selection in inferable menus](https://broadworkbench.atlassian.net/browse/SCP-5325).

This satisfies SCP-5243.
